packages/vm/src/imports.rs

```diff
@@ -317,30 +317,38 @@ pub fn do_next<S: Storage, Q: Querier>(
     key_ptr: u32,
     value_ptr: u32,
 ) -> i32 {
-    let item =
-        match with_iterator_from_context::<S, Q, _, _>(ctx, iterator_id, |iter| Ok(iter.next())) {
-            Ok(i) => i,
-            Err(VmError::UninitializedContextData { .. }) => return errors::NO_CONTEXT_DATA,
-            Err(_) => return errors::next::INVALID_ITERATOR,
-        };
+    let item = match with_iterator_from_context::<S, Q, _, _>(ctx, iterator_id, |iter| {
+        Ok(iter.next())
+    }) {
+        Ok(i) => i,
+        Err(VmError::IteratorDoesNotExist { .. }) => return errors::next::ITERATOR_DOES_NOT_EXIST,
+        Err(VmError::UninitializedContextData { .. }) => return errors::NO_CONTEXT_DATA,
+        Err(_) => return errors::next::UNKNOWN,
+    };
 
-    // prepare return values
+    // Prepare return values. Both key and value are Options and will be written if set.
     let (key, value) = match item {
-        Some(Ok(item)) => item,
+        Some(Ok(item)) => (Some(item.0), Some(item.1)),
         Some(Err(_)) => return errors::next::UNKNOWN,
-        None => return errors::NONE, // Return early without writing key. Empty key will later be treated as _no more element_.
+        None => (Some(Vec::<u8>::new()), None), // Empty key will later be treated as _no more element_.
```

Comment:
Okay, so we explicitly zero out the key passed in here, rather than assuming they passed in an empty buffer. That is nice.

Reply:
Jap. That suddenly became relevant because regions are reused in the test code. I don't know if that happens in a runtime flow as well. But I think it is good to not assume the region we get is already `.length = 0`.